File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/part/WirelessEnergyNetworkTerminalPartMachineBase.java
package com.gtladd.gtladditions.common.machine.muiltblock.part;

import com.gregtechceu.gtceu.api.capability.recipe.IO;
import com.gregtechceu.gtceu.api.gui.fancy.IFancyUIProvider;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.feature.IInteractedMachine;
import com.gregtechceu.gtceu.api.machine.feature.IMachineLife;
import com.gregtechceu.gtceu.api.machine.feature.IRecipeLogicMachine;
import com.gregtechceu.gtceu.api.machine.feature.multiblock.IMultiController;
import com.gregtechceu.gtceu.api.machine.multiblock.part.MultiblockPartMachine;
import com.gregtechceu.gtceu.common.data.GTItems;

import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted;
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;

import com.gtladd.gtladditions.common.machine.trait.NetworkEnergyContainer;
import com.hepdd.gtmthings.api.capability.IBindable;
import com.hepdd.gtmthings.utils.TeamUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;

public class WirelessEnergyNetworkTerminalPartMachineBase extends MultiblockPartMachine implements IInteractedMachine, IBindable, IMachineLife {

    protected static final ManagedFieldHolder MANAGED_FIELD_HOLDER = new ManagedFieldHolder(
            WirelessEnergyNetworkTerminalPartMachineBase.class, MultiblockPartMachine.MANAGED_FIELD_HOLDER);

    @Override
    public @NotNull ManagedFieldHolder getFieldHolder() {
        return MANAGED_FIELD_HOLDER;
    }

    @Persisted
    protected UUID uuid;

    @Persisted
    public final NetworkEnergyContainer energyContainer;

    protected final IO io;

    public WirelessEnergyNetworkTerminalPartMachineBase(IMachineBlockEntity holder, IO io) {
        super(holder);
        this.io = io;
        this.energyContainer = createEnergyContainer();
    }

    @Override
    public InteractionResult onUse(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
        if (player.getItemInHand(hand).is(GTItems.TOOL_DATA_STICK.asItem())) {
            setUUID(player.getUUID());
            if (isRemote()) player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_energy_hatch.tooltip.bind", TeamUtil.GetName(player)));
            return InteractionResult.sidedSuccess(isRemote());
        }
        return InteractionResult.PASS;
    }

    @Override
    public boolean onLeftClick(Player player, Level world, InteractionHand hand, BlockPos pos, Direction direction) {
        ItemStack is = player.getItemInHand(hand);
        if (is.isEmpty()) return false;
        if (is.is(GTItems.TOOL_DATA_STICK.asItem())) {
            this.uuid = null;
            if (isRemote()) {
                player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_energy_hatch.tooltip.unbind"));
            }
            return true;
        }
        return false;
    }

    @Override
    public boolean shouldOpenUI(Player player, InteractionHand hand, BlockHitResult hit) {
        return false;
    }

    @Override
    public void onMachinePlaced(@Nullable LivingEntity player, ItemStack stack) {
        if (player != null) this.setUUID(player.getUUID());
    }

    @Override
    public boolean canShared() {
        return false;
    }

    @Nullable
    @Override
    public IFancyUIProvider.@Nullable PageGroupingData getPageGroupingData() {
        return switch (this.io) {
            case IN -> new PageGroupingData("gtceu.multiblock.page_switcher.io.import", 1);
            case OUT -> new PageGroupingData("gtceu.multiblock.page_switcher.io.export", 2);
            case BOTH -> new PageGroupingData("gtceu.multiblock.page_switcher.io.both", 3);
            case NONE -> null;
        };
    }

    protected NetworkEnergyContainer createEnergyContainer(Object... args) {
        return new NetworkEnergyContainer(this, io);
    }

    @Override
    public UUID getUUID() {
        return uuid;
    }

    @Override
    public void setUUID(UUID uuid) {
        this.uuid = uuid;
        for (IMultiController controller : getControllers()) {
            if (controller instanceof IRecipeLogicMachine machine) {
                machine.getRecipeLogic().updateTickSubscription();
            }
        }
    }
}

File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/part/WirelessEnergyNetworkTerminalPartMachineBase.kt
package com.gtladd.gtladditions.common.machine.muiltblock.part

import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.gui.fancy.IFancyUIProvider.PageGroupingData
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.feature.IInteractedMachine
import com.gregtechceu.gtceu.api.machine.feature.IMachineLife
import com.gregtechceu.gtceu.api.machine.feature.IRecipeLogicMachine
import com.gregtechceu.gtceu.api.machine.multiblock.part.MultiblockPartMachine
import com.gregtechceu.gtceu.common.data.GTItems
import com.gtladd.gtladditions.common.machine.trait.NetworkEnergyContainer
import com.hepdd.gtmthings.api.capability.IBindable
import com.hepdd.gtmthings.utils.TeamUtil
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult
import java.util.*

open class WirelessEnergyNetworkTerminalPartMachineBase(holder: IMachineBlockEntity, protected val io: IO) :
    MultiblockPartMachine(holder), IInteractedMachine, IBindable, IMachineLife {
    @field:Persisted
    protected var uuid: UUID? = null
    @field:Persisted
    val energyContainer: NetworkEnergyContainer

    init {
        this.energyContainer = createEnergyContainer()
    }

    override fun onUse(
        state: BlockState?,
        world: Level?,
        pos: BlockPos?,
        player: Player,
        hand: InteractionHand,
        hit: BlockHitResult?
    ): InteractionResult {
        if (player.getItemInHand(hand).`is`(GTItems.TOOL_DATA_STICK.asItem())) {
            setUUID(player.getUUID())
            if (isRemote) player.sendSystemMessage(
                Component.translatable(
                    "gtmthings.machine.wireless_energy_hatch.tooltip.bind",
                    TeamUtil.GetName(player)
                )
            )
            return InteractionResult.sidedSuccess(isRemote)
        }
        return InteractionResult.PASS
    }

    override fun onLeftClick(
        player: Player,
        world: Level?,
        hand: InteractionHand,
        pos: BlockPos?,
        direction: Direction?
    ): Boolean {
        val `is` = player.getItemInHand(hand)
        if (`is`.isEmpty) return false
        if (`is`.`is`(GTItems.TOOL_DATA_STICK.asItem())) {
            this.uuid = null
            if (isRemote) {
                player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_energy_hatch.tooltip.unbind"))
            }
            return true
        }
        return false
    }

    override fun onMachinePlaced(player: LivingEntity?, stack: ItemStack?) {
        if (player != null) this.setUUID(player.getUUID())
    }

    override fun getPageGroupingData(): PageGroupingData? {
        return when (this.io) {
            IO.IN -> PageGroupingData("gtceu.multiblock.page_switcher.io.import", 1)
            IO.OUT -> PageGroupingData("gtceu.multiblock.page_switcher.io.export", 2)
            IO.BOTH -> PageGroupingData("gtceu.multiblock.page_switcher.io.both", 3)
            IO.NONE -> null
        }
    }

    protected fun createEnergyContainer(vararg args: Any?): NetworkEnergyContainer {
        return NetworkEnergyContainer(this, io)
    }

    override fun getUUID(): UUID? {
        return uuid
    }

    override fun setUUID(uuid: UUID?) {
        this.uuid = uuid
        for (controller in controllers) {
            if (controller is IRecipeLogicMachine) {
                controller.recipeLogic.updateTickSubscription()
            }
        }
    }

    override fun shouldOpenUI(player: Player?, hand: InteractionHand?, hit: BlockHitResult?): Boolean = false

    override fun canShared(): Boolean = false

    override fun getFieldHolder(): ManagedFieldHolder = MANAGED_FIELD_HOLDER

    companion object {
        protected val MANAGED_FIELD_HOLDER: ManagedFieldHolder = ManagedFieldHolder(
            WirelessEnergyNetworkTerminalPartMachineBase::class.java, MultiblockPartMachine.MANAGED_FIELD_HOLDER
        )
    }
}
